api/src/jmh/java/io/opentelemetry/trace/propagation/BinaryTraceContextToByteArrayBenchmark.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.trace.propagation;
+
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.SpanId;
+import io.opentelemetry.trace.TraceFlags;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.TraceState;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Warmup;
+
+@State(Scope.Thread)
+public class BinaryTraceContextToByteArrayBenchmark {
+
+  private static final byte SAMPLED_TRACE_OPTIONS_BYTES = 1;
+  private final BinaryTraceContext binaryTraceContext = new BinaryTraceContext();
+  private final TraceFlags sampledTraceOptions = TraceFlags.fromByte(SAMPLED_TRACE_OPTIONS_BYTES);
+  private final TraceState traceStateDefault = TraceState.builder().build();
+  private final List<SpanContext> spanContextList =
+      Arrays.asList(
+          createSpanContext("905734c59b913b4a905734c59b913b4a", "9909983295041501"),
+          createSpanContext("21196a77f299580e21196a77f299580e", "993a97ee3691eb26"),
+          createSpanContext("2e7d0ad2390617702e7d0ad239061770", "d49582a2de984b86"),
+          createSpanContext("905734c59b913b4a905734c59b913b4a", "776ff807b787538a"),
+          createSpanContext("68ec932c33b3f2ee68ec932c33b3f2ee", "68ec932c33b3f2ee"));
+  private SpanContext spanContext = spanContextList.get(0);
+  private Integer iteration = 0;
+
+  @Benchmark
+  @BenchmarkMode({Mode.Throughput, Mode.AverageTime})
+  @Fork(1)
+  @Warmup(iterations = 5, time = 1)
+  @Measurement(iterations = 50_000, time = 1, timeUnit = TimeUnit.MILLISECONDS)
+  public byte[] measureToByteArray() {
+    return binaryTraceContext.toByteArray(spanContext);
+  }
+
+  @TearDown(Level.Iteration)
+  public void tearDown() {
+    this.spanContext = spanContextList.get(++iteration % spanContextList.size());
```

Comment:
Why switch each iteration?

Reply:
Because we wanna test all pairs in benchmark and not to do loop in benchmark

Reply:
Do you expect them to behave differently? If that is the case using params would make it easier to detect a change.
If no difference why not just loop over them?

Reply:
we discussed it with @carlosalberto, @bogdandrutu and @jkwatson that we should have some set of different traceId\spans and get medium time of it
No loop because in jmh we should not create a usual for\while loop directly in  @Benchmark method

Reply:
I think it make sense to have different value for sanity, but this way if you they for some strange reason behave a bit differently (doesn't take much) you will get a score that goes up and down between iterations that will be hard to explain and understand

I would recommend iterating across all of them in the benchmark method. If you want to minimize impact of iteration, store them as an array instead of a list and to a classic for-loop over it.

if you still want the score to be per toByteArray call you can use the OperationsPerInvocation annotation and set the value to the size of the array.

Reply:
@sfriberg my opinion on loops in jmh based on this example it include loop and OperationsPerInvocation https://hg.openjdk.java.net/code-tools/jmh/file/1a48e97d9dea/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_11_Loops.java
but i don't have a lot of experience with jmh. 
Maybe it can be better to do smth like this
```
  @Benchmark
  @OperationsPerInvocation(3)
  public void measureToByteArray() {
    BlackHole.consume(binaryTraceContext.fromByteArray(firstByteArray));
    BlackHole.consume(binaryTraceContext.fromByteArray(secondByteArray));
    BlackHole.consume(binaryTraceContext.fromByteArray(thirdByteArray));
  }

```

Reply:
Yeah, generally it is nice to avoid loops and but in cases like this is it nice to iterate over the entries.

I see two options to consider
For the first one you need to consider the overhead of the module calculation compared to the rest of the benchmarked code to ensure that doesn't end up being the main part of your execution
```
  private static final byte[] bytes = ...;
  private int index = 0;
  @Benchmark
  public byte[] measureToByteArray() {
    b = binaryTraceContext.fromByteArray(bytes[index]));
    index = (index + 1) % bytes.length
    return b;
  }
```

To ensure a side effect you can also do this basic add in this case which should be trivial enough to not really impact your benchmark much and probably generate less code than the blackhole call.
```
  private static final int COUNT = X;
  @Benchmark
  @OperationsPerInvocation(COUNT)
  public int measureToByteArray() {
    int result = 0;
    for (int = 0; i < COUNT; i++) {
        result = binaryTraceContext.fromByteArray(bytes[i]).length;
    }
    return result;
  }
```

Reply:
good idea i will try to rewrite Http and Binary benchmarks